### PR TITLE
Adjoint-mode for the LSS method

### PR DESCRIPTION
**Pull request by frankschae**
Implements the adjoint mode wrt #420 . Parameter-dependence in the loss function is now explicitly tested as well. I am not sure if windowing would work the same way, so I implemented only the time-dilation option. 

(see related issue: #102)

### CODE REVIEW

File: src/lss.jl [hunk 1]
```diff
@@ -427,3 +427,158 @@ function __solve(prob::ForwardLSSProblem,sensealg::ForwardLSS,alpha::Cos2Windowi
...
+  __solve(prob,prob.sensealg,prob.sensealg.alpha,t0skip,t1skip)
+end
+
+function __solve(prob::AdjointLSSProblem,sensealg::AdjointLSS,alpha::Number,t0skip,t1skip)
+  @unpack sol, S, Δt, diffcache, h, b, wa, res, g, g0, umid = prob
+  @unpack wBinv, B, E, Smat = S
+  @unpack dg_val, pgpp, pgpp_config, numparams, numindvar, uf, f, f_cache, pJ, pf, paramjac_config = diffcache
+
+  b .= E*h + B*wBinv
+  wa .= Smat\b
```
Comment (ChrisRackauckas) [inline src/lss.jl:552]: pre-caching the Smat factorization here?

Comment (frankschae) [reply src/lss.jl:552]: Should this go to `AdjointLSSProblem` then? In this case `Smat\b` is only called once.

Comment (ChrisRackauckas) [reply src/lss.jl:552]: yeah, just `lu!(Smat)` before saving it.

**Review by ChrisRackauckas — APPROVED** [review]

File: src/lss.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/lss.jl]: is this the same as forward LSS? If it's not, we should probably get that updated. I forgot about `lu!` for that 😅

**Comment by ChrisRackauckas** [comment]
Shadowing could be a good use case for DiffEqGPU?

File: src/lss.jl (hunk 1 above)
Comment (frankschae) [reply src/lss.jl]: yup I updated it for both. `lu!(Smat)` just returned the `LU` decomposition though (i.e., `Smat` stayed the same), so I went for removing `Smat` from the `LSSSchur` struct and instead added a field storing the LU decomposition to the respective LSSProblem.

**Comment by frankschae** [comment]
yeah I think it could benefit from GPUs - at least it's full of matrix-matrix or matrix-vector multiplications.

**Pull request merged**